misk/src/main/kotlin/misk/perf/PauseDetector.kt

```diff
@@ -0,0 +1,140 @@
+package misk.perf
+
+import com.google.common.base.Ticker
+import com.google.common.util.concurrent.AbstractExecutionThreadService
+import io.prometheus.client.Summary
+import misk.concurrent.Sleeper
+import misk.metrics.v2.Metrics
+import misk.metrics.v2.PeakGauge
+import org.slf4j.event.Level
+import wisp.logging.getLogger
+import java.time.Duration
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import java.util.concurrent.TimeUnit.NANOSECONDS
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Detects and records pauses experienced by the VM. Garbage collection is a common source of
+ * pauses, but pauses can be caused by many things: CPU throttling, excessive swapping, excessive
+ * active threads, etc.
+ *
+ * Heavily inspired by [jhiccup](https://www.azulsystems.com/jHiccup) which is
+ * [public domain](https://creativecommons.org/publicdomain/zero/1.0/)
+ */
+@Singleton
+internal class PauseDetector @Inject constructor(
+  private val config: PauseDetectorConfig,
+  private val ticker: Ticker,
+  private val sleeper: Sleeper,
+  val metrics: Metrics,
+) : AbstractExecutionThreadService() {
+
+  /** Log levels by pause time sorted by severity descending*/
+  private val logLevels: List<LogLevels> = listOf(
+    LogLevels(config.logErrorMillis, Level.ERROR),
+    LogLevels(config.logWarnMillis, Level.WARN),
+    LogLevels(config.logInfoMillis, Level.INFO)
+  )
+
+  /**
+   * Tracks the distribution and total amount of pause time observed.
+   *
+   * (Using a summary instead of a histogram to prefer more accurate node-local data over
+   * more accurate aggregations)
+   */
+  private val pauseSummary: Summary =
+    metrics.summary("jvm_pause_time_summary_ms", "Summary in ms of pause time", listOf())
+
+  /** Tracks peak pause time. */
+  private val pausePeak: PeakGauge =
+    metrics.peakGauge("jvm_pause_time_peak_ms", "Peak gauge of pause time", listOf())
+
+  // No synchronization is necessary for these variables: they are only ever accessed by the
+  // detector thread itself OR by a test harness thread.
+  //
+  // (They are also initialized in the constructor, but that initialization has a happens-before
+  // relationship to the [Thread.start] that starts the thread).
+  /** Used to prevent generating early noise (for example in the first few seconds of startup) */
+  private var shortestObservedDeltaTimeNsec = Long.MAX_VALUE
+  /** Tracks the start time of the last invocation to [sleep] */
+  private var startTime: Long = 0
+
+  override fun run() {
+    while (isRunning) {
```

Comment:
@swankjesse @adrw this doesn't do what I expected..I was under the impression that `AbstractExecutionThreadService` could manage my running lifecycle for me such that all my code needs to do is check `isRunning` on each iteration. However, this fails checks for actually stopping the service when I try to wire this up in skim (see https://github.com/squareup/cash-skim/pull/1959)

I see some other uses of this abstract class implement their own `isRunning` flag as a volatile/AtomicBoolean...is that what I should do here? I feel like I must be missing the trick since I'd expect to have mandatory interfaces to implement if we have to handle the running state ourselves...

Reply:
How long does it sleep for? Is it just that `pauseMillisDouble` is longer than Skim’s tolerance for the transition from STOPPING to STOPPED ?

Reply:
Turns out the issue was the misk testing module installs the `FakeSleeper` by default which makes any sleep calls block indefinitely unless the clock is manually managed. (In contrast, the `FakeTicker`'s implementation of sleep just moves the clock forward).

PR to address here: https://github.com/cashapp/misk/pull/2747